Dev integration test

dai65527 opened the pull request:
テスト周り整理しました。ご確認をお願いいたします！

- 単体テストをまとめて実行できるようにしました
- rubyによる統合テストを作りました
- リポジトリのルートから`make test`で統合テストと単体テストが走るように設定しました。
- github actionsにより、プルリクエスト時に自動で`make test`を走らせるように設定しました（できてるといい...）

### 単体テストまとめ
- /test/unit_test内にMakefile作成し、各ディレクトリ内でmakeを実行するようにしました。

### 統合テスト
- rubyで統合テストを書きました。
- 内容
  - 任意のリクエストを投げて、レスポンスコード、ボディ、ヘッダを確認できるようにしています。
  - 内容が不足していたら都度追加して行きましょう！
- main.rbに書き足していくイメージです。
  - 現状は代表的ケース(GET /)などしか入れていません
  - 書き方の例は書いていますが、わからなかったら教えてください。（追記します）

### `make test`で統合テストと単体テストが走る
- 単体テストと統合テストを走らせるためのシェルスクリプトを作成。
- それをリポジトリルートからmake testすれば走るように変更しました。

### Github Actions
- プルリクエストを出したときに、githubのホストするmacosで`make test` が走るように設定しました。
- まだGitHubで動いていないのでうまくいくかは神のみぞ知る...

dai65527 committed:
add ft_tester (not passed yet)

dai65527 committed:
add test/integration_test/main.rb, WebservTest.rb

frame done

dai65527 committed:
Merge branch 'main' of https://github.com/dai65527/webserv into dev_integration_test

dai65527 committed:
resolve conflict

dai65527 committed:
add tests

- nomal
- location
- autoindex

dai65527 committed:
mod: unit test

update with latest 仕様

dai65527 committed:
update integration_test

- add cases

dai65527 committed:
add script for test

- `make test` from root of repo
- `make test` from /test/unit_test/my_tester/run_test.sh to run integration test
- /test/integration
- /test/run_test.sh to run both intgration and unit tests

dai65527 committed:
add github actions workflows

to run `make test`

dai65527 committed:
mod name of yaml file of github actions

TrainCoffeeBoy commented on Request.cpp:
```diff
@@ -105,6 +105,7 @@ int Request::receive(int sock_fd, Session& session) {
   } else if (ret == 0) {  // if ret == 0, the socket_is closed
     return REQ_CLOSE_CON;
   }
+  write(1, read_buf, ret);
```
ここ、デグレしてますね。削除お願いします。
（conflictとして検出されないのはなぜだろう...)

TrainCoffeeBoy commented on test/integration_test/my_tester/WebservTest.rb:
```diff
@@ -0,0 +1,89 @@
+    end
+
+    # expected results
+    @expectedCode = "200"
```
私、初めてのrubyです...
@expectedCodeだけは"200"でおいてますが、他のexpected resultsと違い処理が不要だからでしょうか？
main.rb側のテストケースを見ていると大丈夫そうですが。

dai65527 committed:
mod my_tester/run_test.sh

- return 1 to exit 1

dai65527 committed:
add ft_tester env

- to test/integration_test/ft_test
- add tester and needed directories and file
- add script run_test.sh to run all tester
- add cases.md not completed yet
